util/include/channels/mpsc_bounded_channel.h

```diff
@@ -0,0 +1,238 @@
+// Concord
+//
+// Copyright (c) 2020 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").
+// You may not use this product except in compliance with the Apache 2.0
+// License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the
+// LICENSE file.
+
+#pragma once
+
+#include <chrono>
+#include <condition_variable>
+#include <memory>
+#include <mutex>
+#include <queue>
+#include <utility>
+
+#include "channel.h"
+
+namespace concord::channel {
+
+// Forward declarations
+template <typename Msg>
+class BoundedMpscSender;
+template <typename Msg>
+class BoundedMpscReceiver;
+
+namespace impl {
+
+// A simple, non-optimized, lock based, bounded MPSC channel.
+//
+// This is an internal type used by the BoundedMpscSender and BoundedMpscReceiver and should not be exposed publicly.
+template <typename Msg>
+class BoundedMpscQueue {
+ public:
+  BoundedMpscQueue(size_t capacity) : capacity_(capacity) {}
+
+ private:
+  template <typename T>
+  friend class ::concord::channel::BoundedMpscSender;
+  template <typename T>
+  friend class ::concord::channel::BoundedMpscReceiver;
+
+  std::optional<size_t> size() {
+    const std::lock_guard<std::mutex> lock(mutex_);
+    return buf_.size();
+  }
+
+  std::optional<size_t> capacity() {
+    const std::lock_guard<std::mutex> lock(mutex_);
+    return capacity_;
+  }
+
+  void addSender() {
+    const std::lock_guard<std::mutex> lock(mutex_);
+    num_senders_++;
+  }
+
+  void removeSender() {
+    bool num_senders = 0;
+    {
+      const std::lock_guard<std::mutex> lock(mutex_);
+      num_senders_--;
+      num_senders = num_senders_;
+    }
+    if (num_senders == 0) {
+      // Signal the receiver to throw a NoSendersError.
+      cv_.notify_one();
+    }
+  }
+
+  void invalidateReceiver() {
+    const std::lock_guard<std::mutex> lock(mutex_);
+    valid_receiver_ = false;
+  }
+
+  size_t capacity_;
+  std::queue<Msg> buf_;
+  mutable std::mutex mutex_;
+  std::condition_variable cv_;
+  size_t num_senders_ = 0;
+  bool valid_receiver_ = true;
```

Comment:
Why not make all simple data members atomic?

Reply:
It doesn't actually work for mutation. You can end up with lost updates that way. With the use of a condition variable, they must all be protected by a mutex. I actually did have num_senders and valid_receiver as atomic, but ended up with lost notifications and a deadlock when I didn't take the lock to mutate. 

This explains it better than I can: https://www.modernescpp.com/index.php/c-core-guidelines-be-aware-of-the-traps-of-condition-variables

Reply:
For capacity, though, since it's fixed, we could avoid taking the lock altogether though I believe.

Reply:
I went ahead and removed the lock from `BoundedMpscQueue::capacity()`